int64 x
int64 y
int64 th
---
int64 v
int64 w